components/layout/table.rs

```diff
@@ -867,3 +966,232 @@ impl<'a> Iterator for TableRowIterator<'a> {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+impl<'table> Iterator for TableCellStyleIterator<'table> {
+    type Item = TableCellStyleInfo<'table>;
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        // FIXME We do this awkward .take() followed by shoving it back in
+        // because without NLL the row_info borrow lasts too long
+        if let Some(mut row_info) = self.row_info.take() {
+            if let Some(rowspan) = row_info.row.incoming_rowspan.get(self.column_index.absolute as usize) {
+                // we are not allowed to use this column as a starting point. Try the next one.
+                if *rowspan > 1 {
+                    self.column_index.advance(1, &self.column_styles);
+                    // put row_info back in
+                    self.row_info = Some(row_info);
+                    // try again
+                    return self.next();
+                }
+            }
+            if let Some(cell) = row_info.cell_iterator.next() {
+                let rowgroup_style = row_info.rowgroup.map(|r| r.style());
+                let row_style = row_info.row.block_flow.fragment.style();
+                let cell = cell.as_table_cell();
+                let (col_style, colgroup_style) = if let Some(column_style)
+                        = self.column_styles.get(self.column_index.relative as usize) {
+                    let styles = (column_style.col_style.clone(), column_style.colgroup_style.clone());
+                    self.column_index.advance(cell.column_span, &self.column_styles);
+
+                    styles
+                } else {
+                    (None, None)
+                };
+                // put row_info back in
+                self.row_info = Some(row_info);
+                return Some(TableCellStyleInfo {
+                    cell,
+                    colgroup_style,
+                    col_style,
+                    rowgroup_style,
+                    row_style,
+                    table_style: self.table_style,
+                })
+            } else {
+                // next row
+                if let Some((group, row)) = self.row_iterator.next() {
+                    self.row_info = Some(TableCellStyleIteratorRowInfo {
+                        row: &row,
+                        rowgroup: group,
+                        cell_iterator: row.block_flow.base.child_iter()
+                    });
+                    self.column_index = Default::default();
+                    self.next()
+                } else {
+                    // out of rows
+                    // row_info stays None
+                    None
+                }
+            }
+        } else {
+            // empty table
+            None
+        }
+    }
+}
+
+impl<'table> TableCellStyleInfo<'table> {
+    fn build_display_list(&self, mut state: &mut DisplayListBuildState) {
+        if !self.cell.visible {
+            return
+        }
```

Comment:
I believe for empty cells (`cell.visible == false`), we should still draw the column and row backgrounds, just not the cell background and contents, per the spec: https://drafts.csswg.org/css2/tables.html#table-layers